### PR TITLE
Update report

### PR DESCRIPTION
**Pull request by pilgrimygy**
PR Checklist

- [ ] Update NEWS.md?

### CODE REVIEW

File: docs/homepage/blog/offline_reinforcement_learning_algorithm_phase1/index.md [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+In offline reinforcement learning tasks, VAE\dcite{DBLP:journals/corr/KingmaW13} is often used to learn from datasets to approximate behavior policy.
+
+VAE we implemented consists of two neural network: `encoder` and `decoder` ([link](https://github.com/JuliaReinforcementLearning/ReinforcementLearning.jl/blob/master/src/ReinforcementLearningCore/src/policies/q_based_policies/learners/approximators/neural_network_approximator.jl)).
+```julia
+Base.@kwdef struct VAE{E, D}
+    encoder::E
+    decoder::D
+end
+```
+In the encoding stage, it accepts input state and action and outputs the mean and standard deviation of the distribution. Afterward, the hidden action is obtained by sampling the distribution. In the decoding stage, state and hidden action are used as the input to reconstruct action.
```
Comment (findmyway) [inline docs/homepage/blog/offline_reinforcement_learning_algorithm_phase1/index.md:128]:
```suggestion
In the encoding stage, it accepts input state and action and outputs the mean and standard deviation of the distribution. Afterward, the hidden action is obtained by sampling from the resulted distribution. In the decoding stage, state and hidden action are used as the input to reconstruct action.
```

File: docs/homepage/blog/offline_reinforcement_learning_algorithm_phase1/index.md [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+
+During training, we call the `vae_loss` function to get the reconstruction loss and KL loss. The specific task determines the ratio of these two losses.
```
Comment (findmyway) [inline docs/homepage/blog/offline_reinforcement_learning_algorithm_phase1/index.md:130]: Add a link to `KL` loss here.

File: docs/homepage/blog/offline_reinforcement_learning_algorithm_phase1/index.md [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+end
+### Decode
+function decode(model::VAE, state, z)
+    ### Omit specific code
+    return a
+end
+```
+
+#### Offline RL Algorithms
+We used the existing algorithms and hooks to train the offline RL algorithm to create datasets in several environments (such as CartPole, Pendulum) for training. This work can guide the subsequent development of package RLDataset.jl, for example:
```
Comment (findmyway) [inline docs/homepage/blog/offline_reinforcement_learning_algorithm_phase1/index.md:155]: Same as above, add links to `CartPole` and `Pendulum` in the docs.

File: docs/homepage/blog/offline_reinforcement_learning_algorithm_phase1/index.md [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+```julia
+gen_dataset("JuliaRL-CartPole-DQN", policy, env)
+```
+
+##### Benchmark
+We implemented and experimented with offline DQN (in discrete action space) and offline SAC (in continuous action space) as benchmarks. The performance of offline DQN in Cartpole environment:
```
Comment (findmyway) [inline docs/homepage/blog/offline_reinforcement_learning_algorithm_phase1/index.md:161]: The tense you used in this report is kind of confusing to me. Sometimes the present tense is used, now the past tense here. Better to unify them all.

File: docs/homepage/blog/offline_reinforcement_learning_algorithm_phase1/index.md [hunk 5]
```diff
@@ -0,0 +1,315 @@
...
+
+Besides, we implement corresponding functions `π`, `update!` and `sample`. `π` is used to select the action, whose form is determined by the type of action space. `update!` can be used in two stages. In `PreExperiment` stage, we can call this function for pre-training algorithms with `pretrain_step` parameters (such as PLAS). In `PreAct` stage, we call this function for training the `learner`. In function `update!`, we need to call function `sample` to sample a batch of data from the dataset. With the development of RLDataset.jl, the `sample` function will be deprecated.
+
+We can quickly call the offline version of the existing algorithms with almost no additional code with this framework. Therefore, the implementation and performance testing of offline DQN and offline SAC can be completed soon. For example:
```
Comment (findmyway) [inline docs/homepage/blog/offline_reinforcement_learning_algorithm_phase1/index.md:67]: I think this is the first place to mention **SAC**, better to add a reference here.

File: docs/homepage/blog/offline_reinforcement_learning_algorithm_phase1/index.md [hunk 6]
```diff
@@ -0,0 +1,315 @@
...
+After adding CQL loss, the performance of offline DQN improve.
+
+\dfig{body;JuliaRL_OfflineDQN_CQL_CartPole.png}
+
+Currently, this function only supports discrete action space and CQL(H) method. 
+
+##### Critic Regularizer Regression (CRR)
+CRR\dcite{DBLP:conf/nips/0001NZMSRSSGHF20} is a Behavior Cloning (BC) based method. To filter out bad actions and enables learning better policies from low-quality data, CRR utilizes the advantage function to regularize the learning objective of the actor. Pseudocode is as follows:
+
+\dfig{body;CRR.png}
```
Comment (findmyway) [inline docs/homepage/blog/offline_reinforcement_learning_algorithm_phase1/index.md:196]: Add a source link of this picture?

File: docs/homepage/blog/offline_reinforcement_learning_algorithm_phase1/index.md [hunk 7]
```diff
@@ -0,0 +1,315 @@
...
+
+\dfig{body;PLAS1.png}
+
+PLAS pre-trains a CVAE (Conditional Variational Auto-Encoder) to constrain policy. In the pre-training phase, PLAS samples state-action pairs to train CVAE. PLAS needs to learn a deterministic policy mapping state to latent action and then uses CVAE mapping latent action to action in the training phase. When PLAS mapping state or latent action, it needs to use `tanh` function to limit the output range. 
+
+The advantage of pre-training VAE is that it can accelerate the convergence, and it is easier to train when encountered with complex action spaces and import existing VAE models. Its pseudocode is as follows:
+
+\dfig{body;PLAS2.png}
+
+Please refer to this link for specific code ([link](https://github.com/JuliaReinforcementLearning/ReinforcementLearning.jl/blob/master/src/ReinforcementLearningZoo/src/algorithms/offline_rl/PLAS.jl)). The brief function parameters are as follows:
```
Comment (findmyway) [inline docs/homepage/blog/offline_reinforcement_learning_algorithm_phase1/index.md:237]: Better to use the link in docs

**Review by findmyway — CHANGES_REQUESTED** [review]

File: docs/homepage/blog/offline_reinforcement_learning_algorithm_phase1/index.md (hunk 7 above)
Comment (pilgrimygy) [reply docs/homepage/blog/offline_reinforcement_learning_algorithm_phase1/index.md:237]: What do the links in the docs? Is it a brief introduction later?

Comment (findmyway) [reply docs/homepage/blog/offline_reinforcement_learning_algorithm_phase1/index.md:237]: I mean this link: https://juliareinforcementlearning.org/docs/rlzoo/#ReinforcementLearningZoo.PLASLearner-Tuple{}

Comment (pilgrimygy) [reply docs/homepage/blog/offline_reinforcement_learning_algorithm_phase1/index.md:237]: Well.

File: docs/homepage/blog/offline_reinforcement_learning_algorithm_phase1/index.md (hunk 6 above)
Comment (pilgrimygy) [reply docs/homepage/blog/offline_reinforcement_learning_algorithm_phase1/index.md:196]: Source link in paper
